eth_account/_utils/structured_data/hashing.py

```diff
@@ -157,111 +144,94 @@ def get_array_dimensions(data):
-            yield item
+def encode_field(types, name, field_type, value):
+    if value is None:
+        raise ValueError(f"Missing value for field {name} of type {field_type}")
 
+    if field_type in types:
+        return ('bytes32', keccak(encode_data(field_type, types, value)))
 
-@to_tuple
-def _encode_data(primary_type, types, data):
-    # Add typehash
-    yield "bytes32", hash_struct_type(primary_type, types)
+    if field_type == "bytes":
+        if not isinstance(value, bytes):
+            raise TypeError(
+                f"Value of field `{name}` ({value}) is of the type `{type(value)}`, "
+                f"but expected bytes value"
+            )
 
-    # Add field contents
-    for field in types[primary_type]:
-        value = data[field["name"]]
-        if field["type"] == "string":
-            if not isinstance(value, str):
-                raise TypeError(
-                    "Value of `{0}` ({2}) in the struct `{1}` is of the type `{3}`, but expected "
-                    "string value".format(
-                        field["name"],
-                        primary_type,
-                        value,
-                        type(value),
-                    )
-                )
-            # Special case where the values need to be keccak hashed before they are encoded
-            hashed_value = keccak(text=value)
-            yield "bytes32", hashed_value
-        elif field["type"] == "bytes":
-            if not isinstance(value, bytes):
+        return ('bytes32', keccak(value))
+
+    if field_type == "string":
+        if not isinstance(value, str):
+            raise TypeError(
+                f"Value of field `{name}` ({value}) is of the type `{type(value)}`, "
+                f"but expected string value"
+            )
+
+        return ('bytes32', keccak(text=value))
+
+    if field_type[-1] == "]":
+        # Get the dimensions from the value
+        array_dimensions = get_array_dimensions(value)
+        # Get the dimensions from what was declared in the schema
+        parsed_field_type = parse(field_type)
+
+        for i in range(len(array_dimensions)):
+            if len(parsed_field_type.arrlist[i]) == 0:
+                # Skip empty or dynamically declared dimensions
+                continue
+            if array_dimensions[i] != parsed_field_type.arrlist[i][0]:
+                # Dimensions should match with declared schema
                 raise TypeError(
-                    "Value of `{0}` ({2}) in the struct `{1}` is of the type `{3}`, but expected "
-                    "bytes value".format(
-                        field["name"],
-                        primary_type,
-                        value,
-                        type(value),
-                    )
+                    f"Array data `{value}` has dimensions `{array_dimensions}`"
+                    f" whereas the schema has dimensions "
+                    f"`{tuple(map(lambda x: x[0], parsed_field_type.arrlist))}`"
```

Comment:
I noticed a few issues here when dynamic arrays are used along with defined length arrays. `x[0]` will throw an "index out of bounds" error since it's an empty tuple for dynamic array dimensions. I'm going to think a bit more on this one as I think it has led to some other test cases I want to test out as well. It would be useful to compare the values accepted here to values accepted in Solidity and make sure they match as I think I'm seeing some discrepancies with multi arrays and validation errors.